src/main/scala/com/microsoft/ml/spark/lightgbm/LightGBMBooster.scala

```diff
@@ -174,6 +206,48 @@ class LightGBMBooster(val model: String) extends Serializable {
     predLeafToArray(leafIndexDataOutPtr)
   }
 
+  protected def shapForCSR(sparseVector: SparseVector): Array[Double] = {
+    val numCols = sparseVector.size
+    val kind = lightgbmlibConstants.C_API_PREDICT_CONTRIB
+
+    val datasetParams = "max_bin=255 is_pre_partition=True"
+    val dataInt32bitType = lightgbmlibConstants.C_API_DTYPE_INT32
+    val data64bitType = lightgbmlibConstants.C_API_DTYPE_FLOAT64
+
+    ensureShapDataCreated()
+
+    LightGBMUtils.validate(
+      lightgbmlib.LGBM_BoosterPredictForCSRSingle(
+        sparseVector.indices, sparseVector.values,
+        sparseVector.numNonzeros,
+        boosterPtr, dataInt32bitType, data64bitType, 2, numCols,
+        kind, -1, datasetParams,
+        shapDataLengthLongPtr, shapDataOutPtr), "Booster Predict")
+
+    shapToArray(shapDataOutPtr)
+  }
+
+  protected def shapForMat(row: Array[Double]): Array[Double] = {
```

Comment:
does using 
```protected def predictScoreForMat(row: Array[Double], kind: Int, classification: Boolean): Array[Double]```
with
```kind = lightgbmlibConstants.C_API_PREDICT_CONTRIB```
not work in the same way here?  what is different about this method or why would it be used instead of predictScoreForMat?

Reply:
same thing with predictScoreForCSR.
I'm trying to understand why we need to define these separate methods with their own shapDataLengthLongPtr and shapDataOutPtr instead of just calling those with a different kind parameter value.

Reply:
Well, 

Actually I think it could be easily refactorizable, it is actually doing something analogous to predictLeafForMat, we could have a generic predictMat and predictCSR method that gets the dataPtr and dataLenghtPtr as arguments. 

But I am not sure we could reuse the same dataOutPtr, although maybe we could use one with a size as big as the one that needs the most, and just keep 3 lengths, one to 1, one to numFeatures and the other to numTotalModel.

As I say, I might be wrong

Reply:
the new changes look much better